README.md
# cljemu

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar cljemu-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2020 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-04-11
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-04-11
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/cljemu/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/cljemu/compare/0.1.0...0.1.1


test/cljemu/core_test.clj
(ns cljemu.core-test
  (:require [clojure.test :refer :all]
            [cljemu.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/cljemu/core.clj
(ns cljemu.core
  (:gen-class)
  (:require [cljfx.api :as fx]
            [cljemu.state :refer :all]
            [cljemu.chip8 :as chip]
            [cljemu.utils :as u])
  (:import [javafx.scene.canvas Canvas]
           [javafx.scene.canvas GraphicsContext]
           [javafx.scene.paint Color]))
(def height 32)
(def width 64)

(def p-size 4)

(def *state (atom (get-initial-state width height)))

(defn draw [^GraphicsContext graphics state x y]
  (doto graphics 
      (.setFill (get-fill *state x y))
      (.fillRect (* x p-size) (* y p-size) p-size p-size)))

(defn draw-canvas [^Canvas canvas state]
  (let [graphics (.getGraphicsContext2D canvas)]
  (loop [i (- (:height @*state) 1)]
    (when (>= i 0)
      (loop [j (- (:width @*state) 1)]
        (when (>= j 0)
          (draw graphics state j i)
          (recur (- j 1))))
    (recur (- i 1))))))

(defn fill [state]
  (doto state
    (swap! assoc :display (set-fill state 0 0 1))
    (swap! assoc :display (set-fill state 0 31 1))))

(defn handle [event]
  (let [st @*state 
        {:keys [event/type]} event]
    (case type
      ::set-fill (fill *state)
      ::clear-fill (chip/cls *state 0x0000))))

(def actual-handler
  (-> handle
      (fx/wrap-co-effects {:state *state})))

(defn canvas-screen [{:keys [state width height]}]
  {:fx/type :canvas
   :width width
   :height height
   :draw (fn [canvas]
           (draw-canvas canvas state))})


(def renderer
  (fx/create-renderer
    :middleware
    (fx/wrap-map-desc
      (fn [canvas-state]
        {:fx/type :stage
         :showing (not (:terminated @*state))
         :title "Emulator"
         :width (+ (* p-size width) 240)
         :height (+ (* p-size height) 240)
         :scene {:fx/type :scene
                 :root {:fx/type :h-box
                        :padding 50
                        :spacing 10
                        :children [{:fx/type :v-box
                                    :children [{:fx/type canvas-screen
                                                :width (* p-size width)
                                                :height (* p-size height)
                                                :value canvas-state
                                                :state canvas-state
                                                :on-value-changed #(reset! *state %)}
                                               {:fx/type :h-box
                                                :children [{:fx/type :button
                                                            :text "pause/unpause"
                                                            :on-action (fn [_] 
                                                                         (swap! *state assoc :running (not (:running @*state))))}
                                                           {:fx/type :button
                                                            :text "Next"
                                                            :on-action (fn [_] (chip/cycle-once *state))}
                                                           {:fx/type :button
                                                            :text "Terminate"
                                                            :on-action (fn [_] (swap! *state assoc :terminated true))}
                                                           {:fx/type :button
                                                            :text "clear"
                                                            :on-action (fn [_] (chip/cls *state 0x0000))}
                                                           {:fx/type :button
                                                            :text "Log"
                                                            :on-action (fn [_] (u/debugging *state "logging" (str "stopping: "(:running @*state))))}]}]}]}}}))))

(defn -main
  [& args]
  (fx/mount-renderer *state renderer)
  (loop [i 0
         j 0]
    (when (< j (count chip/sprites))
      (u/update-memory *state i (nth chip/sprites j))
      (recur (inc i) (inc j))))
  (u/info "Done loading sprites")
  ;(u/load-rom *state "./roms/test_opcode.ch8")
  ;(u/load-rom *state "./roms/zero_demo.ch8")
  ;(u/load-rom *state "./roms/particle_demo.ch8")
  ;(u/load-rom *state "./roms/tetris.ch8")
  ;(u/load-rom *state "./roms/clock_program.ch8")
  (u/load-rom *state "./roms/maze_demo.ch8")
  (u/load-rom *state "./roms/Airplane.ch8")
  ;(u/load-rom *state "./roms/stars.ch8")
  ;(u/load-rom *state "./roms/hires.ch8")
  ;(u/load-rom *state "./roms/BC_test.ch8")
  (swap! *state assoc :pc 0x200)
  (while (not (:terminated @*state)) 
    (when (:running @*state)
      (chip/cycle-once *state))
      (Thread/sleep 7))
  (u/info "Terminating! Goodbye!"))



src/cljemu/state.clj
(ns cljemu.state
  (:require [clojure.tools.logging :as log]
            [cljemu.utils :as u])
  (:import  [javafx.scene.paint Color]))


(defn random-set [arr]
  (let [a2 arr
        n (count arr)]
    (doseq [i (range (inc n))]
      (aset a2 i 1))
    a2))

(def key-count 16)

(def register-count 16)

(def memory-size 4096)

(defn get-initial-state [width height]
  {:display (int-array (* width height))
   :width width
   :height height
   :debug true
   :running true
   :terminated false
   :trace []
   :sprites_start 0x050
   :key_count key-count
   :key_input (u/empty-vec key-count)
   :memory (u/empty-vec memory-size)
   :gpio (u/empty-vec register-count)
   :stack (u/empty-vec 512) ; not sure how big the stack needs to be
   :stack-pointer 0
   :sound_timer 0
   :delay_timer 0
   :index 0
   :loaded-rom false
   :draw-flag false
   :rom ""
   :pc 0})

(defn position-to-index [state x y]
  (+ (* y (:width @state)) x))

(defn get-val [state x y]
  (nth (:display @state) 
                (position-to-index state x y)))
(defn get-fill [state x y]
  (if (= 1 (get-val state x y))
    Color/WHITE
    Color/BLACK))

(defn set-fill 
  ([state x y i]
   (let [display (:display @state)]
     (aset display (position-to-index state x y) i)
     (swap! state assoc :display display)))
  ([state x y] (set-fill state x y 1)))


(defn clear-fill [state x y]
  (set-fill state x y 0))

(defn get-display [state]
  (when (:debug @state) (log/info (into [] (:display @state)))))



src/cljemu/utils.clj
(ns cljemu.utils
  (:require [clojure.tools.logging :as log]))

(defn log-i [v]
  (log/info v)
  v)

(defn- func-name [f] 
  (clojure.string/replace (second (re-find #"^.+\$(.+)\@.+$" (str f))) #"\_QMARK\_" "?"))

(defn debug [args]
  (when false
    (log/info args)))

(defn info [args]
    (log/info args))

(defn debugging [state f msg]
  (log/info (str "(" f "): " msg " @ " 
         " pc: " (:pc @state)
         " index: " (:index @state)
         " gpio: " (into [] (:gpio @state))
         " trace: " (into [] (:trace @state))
         " memory: " (into [] (:memory @state)))))

(defn error [state f msg]
  (log/error 
    (str "(" f "): " msg " @ " 
         " pc: " (:pc @state)
         " index: " (:index @state)
         " gpio: " (into [] (:gpio @state))
         " trace: " (into [] (:trace @state))
         " memory: " (into [] (:memory @state)))))

(defn peek-stack [state]
  (peek (:stack @state)))

(defn pop-stack [state]
  (swap! state assoc :stack (pop (:stack @state))))

(defn push-stack [state v]
  (swap! state assoc :stack (conj (:stack @state) v)))

(defn read-reg [state index]
  (if (> index (count (:gpio @state))) 
    (error state "read-reg" (str "Reading register out of bound " index))
    (bit-and 0xFFFF (nth (:gpio @state) index))))

(defn update-vec-index [vect index v]
  (vec 
    (concat
      (subvec vect 0 index)
      (vec [(bit-and 0xff v)])
      (subvec vect (inc index)))))

(defn update-reg [state index v]
  (swap! state assoc :gpio (update-vec-index (:gpio @state) index v)))

(defn update-memory [state index v]
  (swap! state assoc :memory (update-vec-index (:memory @state) index v))
  state)
  
(defn read-memory [state i]
  (nth (:memory @state) i))

(defn read-input [state k]
  (when (> k (:key_count @state))
    (error state "read-input" (str "invalid key value " k)))
  (nth (:key_input @state) k))


(defn setf [state v] 
  (update-reg state 15 (bit-and 0x1 v)))


(defn empty-vec [l]
  (let [v (transient [])]
  (doseq [i (range 0 l)]
     (conj! v 0))
  (persistent! v))) 

(defn slurp-bytes
  "Slurp the bytes from a slurpable thing"
  [x]
  (with-open [out (java.io.ByteArrayOutputStream.)]
    (clojure.java.io/copy (clojure.java.io/input-stream x) out)
    (.toByteArray out)))

(defn read-rom-to-memory [state rom i]
  (if (< i (count rom))
    (read-rom-to-memory (update-memory state (+ 0x200 i) (nth rom i)) rom (inc i))
    state))
(defn increment-pc [state v]
  (swap! state assoc :pc (+ (:pc @state) v)))
(defn skip-ins [state] 
  (increment-pc state 4))
(defn inc-pc [state]
  (increment-pc state 2))

(defn load-rom [state path]
  (let [rom (slurp-bytes path)]
    (doto state
      (swap! assoc :rom rom)
      (read-rom-to-memory rom 0))))
  
    





src/cljemu/chip8.clj
(ns cljemu.chip8
  (:require [cljemu.utils :as u]
            [cljemu.state :as s]))

(def sprites
  [0xF0 0x90 0x90 0x90 0xF0; 0
   0x20 0x60 0x20 0x20 0x70; 1
   0xF0 0x10 0xF0 0x80 0xF0; 2
   0xF0 0x10 0xF0 0x10 0xF0; 3
   0x90 0x90 0xF0 0x10 0x10; 4
   0xF0 0x80 0xF0 0x10 0xF0; 5
   0xF0 0x80 0xF0 0x90 0xF0; 6
   0xF0 0x10 0x20 0x40 0x40; 7
   0xF0 0x90 0xF0 0x90 0xF0; 8
   0xF0 0x90 0xF0 0x10 0xF0; 9
   0xF0 0x90 0xF0 0x90 0x90; A
   0xE0 0x90 0xE0 0x90 0xE0; B
   0xF0 0x80 0x80 0x80 0xF0; C
   0xE0 0x90 0x90 0x90 0xE0; D
   0xF0 0x80 0xF0 0x80 0xF0; E
   0xF0 0x80 0xF0 0x80 0x80]); F
(defn VX [opcode]
  (bit-and 0x000F (bit-shift-right (bit-and opcode 0x0F00) 8)))

(defn VY [opcode]
  (bit-and 0x000F (bit-shift-right (bit-and opcode 0x00F0) 4)))

(defn cls [state opcode] ; OK
  "Clear the display."
  (u/debug (format "cls %x" opcode))
  (swap! state assoc :display (int-array (* (:width @state) (:height @state))))
  (u/inc-pc state))

(defn ret [state opcode]
    "00EE - RET"
    "Return from a subroutine."
    "The interpreter sets the program counter to the address at the top of the stack, then subtracts 1 from the stack pointer."
    (u/debug (format "ret %x" opcode))
      (swap! state assoc :pc (u/peek-stack state))
      (u/pop-stack state)
      (swap! state assoc :draw-flag true))

(defn jmp [state opcode]
  "1nnn - JP addr
  Jump to location nnn.

  The interpreter sets the program counter to nnn."
  (u/debug (format "jmp %x" opcode))
  (let [nnn (bit-and opcode 0x0fff)]
  (swap! state assoc :pc nnn)))

(defn call-addr [state opcode]
  "2nnn - CALL addr
  Call subroutine at nnn.

  The interpreter increments the stack pointer, then puts the current PC on the top of the stack. The PC is then set to nnn."
  (u/debug (format "call-addr %x" opcode))
  (let [nnn (bit-and opcode 0x0FFF)]
    (u/inc-pc state)
    (u/push-stack state (:pc @state))
    (swap! state assoc :pc nnn)))


(defn se [state opcode]
  "3xkk - SE Vx, byte
  Skip next instruction if Vx = kk.
  The interpreter compares register Vx to kk, and if they are equal, increments the program counter by 2."
  (u/debug (format "se %x" opcode))
  (let [Vx (VX opcode)
        kk (bit-and opcode 0x00ff)]
    (when (= (u/read-reg state Vx) kk) ; skip if equal
      (u/inc-pc state))
      (u/inc-pc state)))


(defn sne [state opcode]
  "4xkk - SNE Vx, byte
  Skip next instruction if Vx != kk.

  The interpreter compares register Vx to kk, and if they are not equal, increments the program counter by 2."
  (u/debug (format "sne %x" opcode))
  (let [Vx (VX opcode)
        kk (bit-and opcode 0x00ff)]
    (when (not (= (u/read-reg state Vx) kk))
      (u/inc-pc state))
      (u/inc-pc state)))


(defn se-reg [state opcode]
  "5xy0 - SE Vx, Vy
  Skip next instruction if Vx = Vy.

  The interpreter compares register Vx to register Vy, and if they are equal, increments the program counter by 2."
  (u/debug (format "se-reg %x" opcode))
  (let [Vx (VX opcode)
        Vy (VY opcode)]
    (when (= (u/read-reg state Vx) (u/read-reg state Vy))
      (u/inc-pc state))
    (u/inc-pc state)))

(defn ld [state opcode]
  "6xkk - LD Vx, byte
  Set Vx = kk.

  The interpreter puts the value kk into register Vx."
  (u/debug (format "ld %x" opcode))
  (let [Vx (VX opcode)
        kk (bit-and opcode 0x00ff)]
    (u/update-reg state Vx kk)
    (u/inc-pc state)))

(defn add [state opcode]
  "7xkk - ADD Vx, byte
  Set Vx = Vx + kk.

  Adds the value kk to the value of register Vx, then stores the result in Vx."
  (u/debug (format "add %x" opcode))
  (let [Vx (VX opcode)
        current-vx (u/read-reg state Vx)
        kk (bit-and opcode 0x00ff)
        result (+ current-vx kk)]
    (u/debug (format "result: %d" result))
    (u/update-reg state Vx (bit-and result 0xff))
    (u/inc-pc state)))

(defn ld-reg [state opcode]
  "8xy0 - LD Vx, Vy
  Set Vx = Vy.

  Stores the value of register Vy in register Vx."
  (u/debug (format "ld-reg %x" opcode))
  (let [Vx (VX opcode)
        Vy (VY opcode)]
    (u/update-reg state Vx (u/read-reg state Vy))
    (u/inc-pc state)))

(defn or-reg [state opcode]
  "8xy1 - OR Vx, Vy
  Set Vx = Vx OR Vy.

  Performs a bitwise OR on the values of Vx and Vy, then stores the result in Vx. 
  A bitwise OR compares the corrseponding bits from two values, and if either bit is 1, then the same bit in the result is also 1. Otherwise, it is 0."
  (u/debug (format "or-reg %x" opcode))
  (let [Vx (VX opcode)
        current-vx (u/read-reg state Vx)
        Vy (VY opcode)
        current-vy (u/read-reg state Vy)]
    (u/update-reg state Vx (bit-or current-vx current-vy))
    (u/inc-pc state)))

(defn and-reg [state opcode]
  "8xy2 - AND Vx, Vy
  Set Vx = Vx AND Vy.

  Performs a bitwise AND on the values of Vx and Vy, then stores the result in Vx. 
  A bitwise AND compares the corrseponding bits from two values, and if both bits are 1, then the same bit in the result is also 1. Otherwise, it is 0."
  (u/debug (format "and-reg %x" opcode))
  (let [Vx (VX opcode)
        current-vx (u/read-reg state Vx)
        Vy (VY opcode)
        current-vy (u/read-reg state Vy)]
    (u/update-reg state Vx (bit-and current-vx current-vy))
    (u/inc-pc state)))

(defn xor-reg [state opcode]
  "8xy3 - XOR Vx, Vy
  Set Vx = Vx XOR Vy.

  Performs a bitwise exclusive OR on the values of Vx and Vy, then stores the result in Vx. 
  An exclusive OR compares the corrseponding bits from two values, and if the bits are not both the same, 
  then the corresponding bit in the result is set to 1. Otherwise, it is 0."
  (u/debug (format "xor-reg %x" opcode))
  (let [Vx (VX opcode)
        current-vx (u/read-reg state Vx)
        Vy (VY opcode)
        current-vy (u/read-reg state Vy)]
    (u/update-reg state Vx (bit-xor current-vx current-vy))
    (u/inc-pc state)))


(defn add-reg [state opcode]
  "8xy4 - ADD Vx, Vy
  Set Vx = Vx + Vy, set VF = carry.

  The values of Vx and Vy are added together. If the result is greater than 8 bits (i.e., > 255,) VF is set to 1, otherwise 0. 
  Only the lowest 8 bits of the result are kept, and stored in Vx."
  (u/debug (format "add-reg %x" opcode))
  (let [Vx (VX opcode)
        current-vx (u/read-reg state Vx)
        Vy (VY opcode)
        current-vy (u/read-reg state Vy)
        sum (+ current-vx current-vy)]
      (if (> sum 255) (u/setf state 1) (u/setf state 0))
      (u/update-reg state Vx (bit-and sum 0xFF))
    (u/inc-pc state)))

(defn sub-reg [state opcode]
  "8xy5 - SUB Vx, Vy
  Set Vx = Vx - Vy, set VF = NOT borrow.

  If Vx > Vy, then VF is set to 1, otherwise 0. Then Vy is subtracted from Vx, and the results stored in Vx."
  (u/debug (format "sub-reg %x" opcode))
  (let [Vx (VX opcode)
        current-vx (u/read-reg state Vx)
        Vy (VY opcode)
        current-vy (u/read-reg state Vy)
        sum (- current-vx current-vy)]
    (if (> current-vx current-vy) (u/setf state 1) (u/setf state 0))
        (u/update-reg state Vx (bit-and sum 0xFF))
        (u/inc-pc state)))

(defn shr [state opcode]
  "8xy6 - SHR Vx {, Vy}
  Set Vx = Vx SHR 1.

  If the least-significant bit of Vx is 1, then VF is set to 1, otherwise 0. Then Vx is divided by 2."
  (u/debug (format "shr %x" opcode))
  (let [Vx (VX opcode)
        current-vx (u/read-reg state Vx)
        quotient (bit-shift-right current-vx 1)]
    (u/setf state (bit-and 0x1 current-vx))
    (u/update-reg state Vx quotient)
    (u/inc-pc state)))

(defn subn [state opcode]
  "8xy7 - SUBN Vx, Vy
  Set Vx = Vy - Vx, set VF = NOT borrow.

  If Vy > Vx, then VF is set to 1, otherwise 0. Then Vx is subtracted from Vy, and the results stored in Vx."
  (u/debug (format "subn %x" opcode))
  (let [Vx (VX opcode)
        current-vx (u/read-reg state Vx)
        Vy (VY opcode)
        current-vy (u/read-reg state Vy)
        diff (- current-vy current-vx)]
    (u/update-reg state Vx diff)
    (if (> current-vy current-vx) (u/setf state 1) (u/setf state 0))
        (u/update-reg state Vx (bit-and 0xFFFF diff))
    (u/inc-pc state)))

(defn shl [state opcode]
  "8xyE - SHL Vx {, Vy}
  Set Vx = Vx SHL 1.

  If the most-significant bit of Vx is 1, then VF is set to 1, otherwise to 0. Then Vx is multiplied by 2."
  (u/debug (format "shl %x" opcode))
  (let [Vx (VX opcode)
        current-vx (u/read-reg state Vx)
        result (bit-shift-left current-vx 1)]
    (u/update-reg state Vx (bit-and 0xFFFF result))
    (u/setf state (bit-and (bit-shift-right current-vx 7) 0x1))
    (u/inc-pc state)))

(defn sne-reg [state opcode]
  "9xy0 - SNE Vx, Vy
  Skip next instruction if Vx != Vy.

  The values of Vx and Vy are compared, and if they are not resequal, the program counter is increased by 2."
  (u/debug (format "sne-reg %x" opcode))
  (let [Vx (VX opcode)
        Vy (VY opcode)]
    (when (not (= (u/read-reg state Vx)
                  (u/read-reg state Vy)))
      (u/inc-pc state))
      (u/inc-pc state)))

(defn ld-i [state opcode]
  "Annn - LD I, addr
  Set I = nnn.

  The value of register I is set to nnn."
  (u/debug (format "ld-i %x" opcode))
  (let [nnn (bit-and opcode 0x0fff)]
    (swap! state assoc :index nnn)
      (u/inc-pc state)))

(defn jp-v0 [state opcode]
  "Bnnn - JP V0, addr
  Jump to location nnn + V0.

  The program counter is set to nnn plus the value of V0."
  (u/debug (format "jp-v0 %x" opcode))
  (let [nnn (bit-and opcode 0x0fff)
        V0 (u/read-reg state 0)]
    (swap! state assoc :pc (+ nnn V0))))

(defn rnd [state opcode]
  "Cxkk - RND Vx, byte
  Set Vx = random byte AND kk.

  The interpreter generates a random number from 0 to 255, which is then ANDed with the value kk. 
  The results are stored in Vx. See instruction 8xy2 for more information on AND."
  (u/debug (format "rnd %x" opcode))
  (let [Vx (VX opcode)
        kk (bit-and opcode 0x00ff)]
    (u/update-reg state Vx (bit-and (rand-int 256) kk))
    (u/inc-pc state)))

(defn drw [state opcode]
  "Dxyn - DRW Vx, Vy, nibble
  Display n-byte sprite starting at memory location I at (Vx, Vy), set VF = collision.

  The interpreter reads n bytes from memory, starting at theresation on XOR, and section 2.4, Display, for more information on the Chip-8 screen and sprites."
  (u/debug (format "drw %x" opcode))
  (let [Vx (VX opcode)
        Vy (VY opcode)
        x (u/read-reg state Vx)
        y (u/read-reg state Vy)
        n (bit-and opcode 0x000f)
        I (:index @state)
        display (:display @state)]
    (u/setf state 0)
    (loop [i 0]
      (when (< i n)
        (let [Byte (u/read-memory state (+ I i))]
          (loop [j 0]
            (when (< j 8)
              ;(u/info (format "x:%d y:%d i:%d j:%d Y:%d X:%d" x y i j (+ y i) (+ x j)))
              (let [Y (mod (+ y i) 32)
                    X (mod (+ x j) 64)]
                (s/set-fill state 
                            X
                            Y
                            (bit-xor 
                              (nth display (s/position-to-index state X Y)) 
                              (bit-and 0x0001 (bit-shift-right Byte (- 8 j)))))
                  (u/setf state (s/get-val state X Y)))
                (recur (inc j)))))
        (recur (inc i))))
    ;(u/info "Finito in drw")
    (u/inc-pc state)))

(defn skp [state opcode]
  "Ex9E - SKP Vx
  Skip next instruction if key with the value of Vx is pressed.

  Checks the keyboard, and if the key corresponding to the value of Vx is currently in the down position, PC is increased by 2."
  (u/debug (format "skp %x" opcode))
  (let [Vx (VX opcode)]
    (when (= (u/read-input state (u/read-reg state Vx)) 1) 
      (u/inc-pc state))
    (u/inc-pc state)))

(defn sknp [state opcode]
  "ExA1 - SKNP Vx
  Skip next instruction if key with the value of Vx is not pressed.

  Checks the keyboard, and if the key corresponding to the value of Vx is currently in the up position, PC is increased by 2."
  (u/debug (format "sknp %x" opcode))
  (let [Vx (VX opcode)]
    (when (= (u/read-input state (u/read-reg state Vx)) 0) 
      (u/inc-pc state))
    (u/inc-pc state)))

(defn ld-dt [state opcode]
  "Fx07 - LD Vx, DT
  Set Vx = delay timer value.

  The value of DT is placed into Vx."
  (u/debug (format "ld-dt %x" opcode))
  (let [Vx (VX opcode)]
    (u/update-reg state Vx (:delay_timer @state))
    (u/inc-pc state)))

(defn ld-k [state opcode]
  "Fx0A - LD Vx, K
  Wait for a key press, store the value of the key in Vx.

  All execution stops until a key is pressed, then the value of that key is stored in Vx."
  (let [Vx (VX opcode)]
        (loop [frst (first (filter #(= (nth (:key_input @state) %1) 1) (doseq [i (range 0 (count (:key_input @state)))])))]
          (when (not (nil? frst))
            (u/update-reg state Vx frst)
            (recur (first (filter #(= (nth (:key_input @state) %1) 1) (doseq [i (range 0 (count (:key_input @state)))]))))))
    (u/inc-pc state)))

(defn ld-dt-reg [state opcode]
  "Fx15 - LD DT, Vx
  Set delay timer = Vx.

  DT is set equal to the value of Vx."
  (u/debug (format "ld-dt-reg %x" opcode))
  (let [Vx (VX opcode)]
    (swap! state assoc :delay_timer (u/read-reg state Vx))
    (u/inc-pc state)))

(defn ld-st-reg [state opcode]
  "Fx18 - LD ST, Vx
  Set sound timer = Vx.

  ST is set equal to the value of Vx."
  (u/debug (format "ld-st-reg %x" opcode))
  (let [Vx (VX opcode)]
    (swap! state assoc :sound_timer (u/read-reg state Vx))
    (u/inc-pc state)))

(defn add-i [state opcode]
  "Fx1E - ADD I, Vx
  Set I = I + Vx.

  The values of I and Vx are added, and the results are stored in I."
  (u/debug (format "add-i %x" opcode))
  (let [Vx (VX opcode)
        I (:index @state)
        current-vx (u/read-reg state Vx)]
    (if (> (+ I current-vx) 0xFFFF) (u/setf state 1) (u/setf state 0))
    (swap! state assoc :index (+ I current-vx))
    (u/inc-pc state)))

(defn ld-sprite [state opcode]
  "Fx29 - LD F, Vx
  Set I = location of sprite for digit Vx.

  The value of I is set to the location for the hexadecimal sprite corresponding to the value of Vx. See section 2.4, Display, for more information on the Chip-8 hexadecimal font."
  (u/debug (format "ld-sprite %x" opcode))
  (let [Vx (VX opcode)
        current-vx (u/read-reg state Vx)
        I (* current-vx 5)]
    (swap! state assoc :index I)
    (u/inc-pc state)))

(defn ld-b [state opcode]
  "Fx33 - LD B, Vx
  Store BCD representation of Vx in memory locations I, I+1, and I+2.

  The interpreter takes the decimal value of Vx, and places the hundreds digit in memory at location in I, the tens digit at location I+1, and the ones digit at location I+2."
  (u/debug (format "ld-b %x" opcode))
  (let [Vx (VX opcode)
        I (:index @state)
        current-vx (u/read-reg state Vx)
        hundreds (int (/ (mod current-vx 1000) 100))
        tens (int (/ (mod current-vx 100) 10))
        ones (int (mod current-vx 10))]
   (u/update-memory state I hundreds) 
   (u/update-memory state (inc I) tens) 
   (u/update-memory state (+ I 2) ones)
   (u/inc-pc state)))

(defn ld-to-memory [state opcode]
  "Fx55 - LD [I], Vx
  Store registers V0 through Vx in memory starting at location I.

  The interpreter copies the values of registers V0 through Vx into memory, starting at the address in I."
  (u/debug (format "ld-to-memory %x" opcode))
  (let [Vx (VX opcode)
        I (:index @state)]
   (loop [i 0]
      (when (<= i Vx)
        (u/update-memory state (+ i I) (u/read-reg state i))
       (recur (inc i))))
   ;(swap! state assoc :index (+ (:index @state) Vx 1))
   (u/inc-pc state)))

(defn ld-from-memory [state opcode]
  "Fx65 - LD Vx, [I]
  Read registers V0 through Vx from memory starting at location I.

  The interpreter reads values from memory starting at location I into registers V0 through Vx."
  (u/debug (format "ld-from-memory %x" opcode))
  (let [Vx (VX opcode)
        I (:index @state)]
    (loop [x 0]
      (when (<= x Vx)
        (u/update-reg state x (u/read-memory state (+ x I)))
        (recur (inc x))))
    ;(swap! state assoc :index (+ (:index @state) Vx 1))
    (u/inc-pc state)))

(def instructions 
  {0x00E0 cls
   0x00EE ret    
   0x1000 jmp
   0x2000 call-addr
   0x3000 se
   0x4000 sne
   0x5000 se-reg
   0x6000 ld
   0x7000 add
   0x8000 ld-reg
   0x8001 or-reg
   0x8002 and-reg
   0x8003 xor-reg
   0x8004 add-reg
   0x8005 sub-reg
   0x8006 shr
   0x8007 subn
   0x800E shl
   0x9000 sne-reg
   0xA000 ld-i
   0xB000 jp-v0
   0xC000 rnd
   0xD000 drw
   0xE00E skp
   0xE0A1 sknp
   0xF007 ld-dt
   0xF00A ld-k
   0xF015 ld-dt-reg
   0xF018 ld-st-reg
   0xF01E add-i
   0xF029 ld-sprite
   0xF033 ld-b
   0xF055 ld-to-memory
   0xF065 ld-from-memory
   })
(defn handle-MSByte-E [opcode state]
  (let [LSByte (bit-and 0x000F opcode)]
    (case LSByte 
      0x000E (get instructions 0xE00E)
      0x0001 (get instructions 0xE0A1)
      (do (u/error state "handle-MSByte-E" (format "Invalid opcode %x" opcode))  (throw (Exception. (format "Invalid opcode (special) %x" opcode)))))))
(defn handle-MSByte-F [opcode state]
  (let [LSByte (bit-and 0x000F opcode)]
    (case LSByte 
      0x0007 (get instructions 0xF007)
      0x000A (get instructions 0xF00A)
      0x0008 (get instructions 0xF018)
      0x000E (get instructions 0xF01E)
      0x0009 (get instructions 0xF029)
      0x0003 (get instructions 0xF033)
      0x0005 (let [Byte (bit-and 0x00FF opcode)]
               (case Byte 
                 0x0055 (get instructions 0xF055)
                 0x0065 (get instructions 0xF065)
                 0x0015 (get instructions 0xF015)
                 (do (u/error state "handle-MSByte-F" (format "Invalid opcode %x" opcode))  (throw (Exception. (format "Invalid opcode (special) %x" opcode))))))
      (do (u/error state "handle-MSByte-F" (format "Invalid opcode %x" opcode))  (throw (Exception. (format "Invalid opcode (special) %x" opcode)))))))


(defn handle-MSByte-8 [opcode state]
  (let [LSByte (bit-and 0x000F opcode)]
    (case LSByte 
      0x0000 (get instructions 0x8000)
      0x0001 (get instructions 0x8001)
      0x0002 (get instructions 0x8002)
      0x0003 (get instructions 0x8003)
      0x0004 (get instructions 0x8004)
      0x0005 (get instructions 0x8005)
      0x0006 (get instructions 0x8006)
      0x0007 (get instructions 0x8007)
      0x000E (get instructions 0x800E)
      (do (u/error state "handle-MSByte-8" (format "Invalid opcode %x" opcode)) (throw (Exception. (format "Invalid opcode (special) %x" opcode)))))))

(defn to-instruction [opcode state]
  ; special case for 0xF0ij 
  (let [MSByte (bit-and (bit-shift-right opcode 12) 0x000F)]
    (case MSByte
      0x0001 (get instructions 0x1000)
      0x0002 (get instructions 0x2000)
      0x0003 (get instructions 0x3000) 
      0x0004 (get instructions 0x4000)
      0x0005 (get instructions 0x5000)
      0x0006 (get instructions 0x6000)
      0x0007 (get instructions 0x7000)
      0x0009 (get instructions 0x9000)
      0x000A (get instructions 0xA000)
      0x000B (get instructions 0xB000)
      0x000C (get instructions 0xC000)
      0x000D (get instructions 0xD000)
      0x0008 (handle-MSByte-8 opcode state)
      0x000E (handle-MSByte-E opcode state)
      0x000F (handle-MSByte-F opcode state)
   (do (u/error state "to-instruction" (format "Invalid opcode %x" opcode)) (throw (Exception. (format "Invalid opcode (special) %x" opcode)))))))

    
(defn perform-opcode [state nibble-one nibble-two]
    (let [opcode (bit-or
                   (bit-and 0xff00 (bit-shift-left nibble-one 8))
                   (bit-and nibble-two 0x00ff))
          instruction (case opcode
                        (0x00E0 0x0EE) (get instructions opcode)
                        (to-instruction opcode state ))]
      (swap! state assoc :trace (conj (:trace @state) (format "%x" opcode)))
      (instruction state opcode)))
  
(defn cycle-once [state]
  (let [nibble-one (nth (:memory @state) (:pc @state))
        nibble-two (nth (:memory @state) (+ (:pc @state) 1))
        new-delay (if (> (:delay_timer @state) 0) (dec (:delay_timer @state)) 0)
        new-sound (if (> (:sound_timer @state) 0) (dec (:sound_timer @state)) 0)]
    (perform-opcode state nibble-one nibble-two)
    (swap! state assoc :delay_timer new-delay)
    (swap! state assoc :sound_timer new-sound)))
